rfcs/2021-10-29-8621-framing-and-codecs-sinks.md

```diff
@@ -0,0 +1,127 @@
+    fn build(&self) -> crate::Result<BoxedSerializer>;
+}
+```
+
+```rust
+pub trait FramingConfig: Debug + DynClone + Send + Sync {
+    /// Builds a framer from this configuration.
+    ///
+    /// Fails if the configuration is invalid.
+    fn build(&self) -> crate::Result<BoxedFramer>;
+}
+```
+
+These can be built and combined to form an `Encoder`:
+
+```rust
+/// An encoder that can encode structured events to byte messages.
+pub struct Encoder {
+    serializer: BoxedSerializer,
+    framer: BoxedFramer,
+}
+```
+
+`Encoder` implements `tokio_util::codec::Encoder<Event>`. Internally, events first go through the `Serializer` which implements `tokio_util::codec::Encoder<Event>` and are then handed over to the `Framer` which implements `tokio_util::codec::Encoder<Bytes>`.
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues after the RFC is approved:
+
+- [ ] Implementation of support structures (`Encoder`, `SerializerConfig`, `FramingConfig`)
+- [ ] Implementation of selected `encoders`/`framers` (e.g. `JSON` and `newline_delimited`)
+- [ ] Implementation of compatibility layer for legacy reshaping via `EncodingConfiguration::apply_rules`
+- [ ] Example integration with first sink, e.g. `socket` or `http` which benefit most from generic encoding/framing options
+- [ ] Subsequent PRs for each integration to a sink
+
+## Surveyed Sinks
+
+Overview for the current state of sinks regarding encoding:
+
+|sink|encoding config|`.apply_rules`|notes|
+|-|-|-|-|
+|`aws_cloudwatch_logs`| `EncodingConfig<Encoding { Text, Json }>` | ✔︎ | Enveloped in `rusoto_logs::InputLogEvent`. `Text` reads message_key()
+|`aws_kinesis_firehose`| `EncodingConfig<Encoding { Text, Json }>` | ✔︎ | Enveloped in `rusoto_firehose::Record` that serializes to base64. `Text` reads `message_key()` | -
+|`aws_kinesis_streams`| `EncodingConfig<Encoding { Text, Json }>` | ✔︎ | Enveloped in `rusoto_kinesis::PutRecordsRequestEntry`. `Text` reads `message_key()`
+|`aws_s3`| `EncodingConfig<StandardEncodings { Text, Json, Ndjson }>` | ✔︎ | Uses `util::{RequestBuilder, Encoder, Compressor}`. `Text` reads `message_key()`
+|`aws_sqs`| `EncodingConfig<Encoding { Text, Json }>` | ✔︎ | Enveloped in `EncodedEvent<SendMessageEntry>`. `Text` reads `message_key()`
+|`azure_blob`| `EncodingConfig<Encoding { Ndjson, Text }>` | ✔︎ | Enveloped in `EncodedEvent<PartitionInnerBuffer>`. `Text` reads `message_key()`
+|`azure_monitor_logs`| `EncodingConfigWithDefault<Encoding { Default }>` | ✔︎ | Serializes to JSON. Enveloped in HTTP request
+|`blackhole`| - | - | -
+|`clickhouse`| `EncodingConfigWithDefault<Encoding { Default }>` | ✔︎ | Serializes to JSON. Enveloped in HTTP request
+|`console`| `EncodingConfig<Encoding { Text, Json }>` | ✔︎ | `Text` reads `message_key()`
+|`datadog`| `EncodingConfigFixed<DatadogLogsJsonEncoding>` | ✔︎ | Doesn't provide options to encode the event payload separately from the protocol
```

Comment:
There is no `datadog` sink. This is describing the `datadog_logs` sink. The `datadog_events` sink is missing, which is one of the more interesting ones. It doesn't allow setting the encoding at all, and uses the `EncodingConfig` to filter out fields that the datadog events api doesn't accept.